### PR TITLE
First stab at documentation.

### PR DESCRIPTION
**Pull request by kilobyte**
No proper build system, etc, yet.  Formatting needs improvements, too (what's our equivalent for .TP?).

@wlemkows: as our doc mistress, could you please review this and educate me wrt preferred build syste/templates/etc for new projects?

---
This change is [Reviewable](https://reviewable.io/reviews/pmem/vmemcache/121)

### CODE REVIEW

File: doc/vmemcache.md [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+be prevented, but until the callback returns, the entry remains available
+for queries.  The thread that triggered the eviction is blocked in the
+meantime.
+
+
+int vmemcache_on_miss(VMEMcache *cache,
+	const void *key, size_t key_size, void *arg);
+
+Called when a *get* query fails, to provide an opportunity to insert the
+missing key.  If the callback returns non-zero, upon return the query will
```
Comment (ldorau) [inline doc/vmemcache.md]: If the callback returns *zero*, ...

**Comment by kilobyte** [comment]
I think, let's put this into the repository as text, then fix up formatting and add a build system later.  Unlike code, documentation can serve its purpose even without meeting any rules.

**Comment by kilobyte** [comment]
Included changes from pending pull requests wrt ENOSPC and EBUSY.

File: doc/vmemcache.md [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+
+##### Use #####
+
+ssize_t vmemcache_get(VMEMcache *cache,
+	const void *key, size_t key_size,
+	void *vbuf, size_t vbufsize, size_t offset, size_t *vsize);
+
+Searches for an entry for the given *key*; it doesn't have to be
+zero-terminated or be text -- any sequence of bytes of length *key_size*
```
Comment (wlemkows) [inline doc/vmemcache.md]: . -> --

File: doc/vmemcache.md [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+is okay.  If found, the entry's value is copied to *vbuf* that has space
+for *vbufsize* bytes, optionally skipping *offset* bytes at the start.
+No matter if the copy was truncated or not, it's true size is stored into
+*vsize*; *vsize* remains unmodified if the key was not found.
+
+Return value is -1 on error, 0 if the key was not found, or number of bytes
+successfully copied otherwise.  Note that 0 may be legitimately returned
```
Comment (wlemkows) [inline doc/vmemcache.md]: double spaces

File: doc/vmemcache.md [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+even if the key was found (reading past end, zero-sized value, etc) --
```
Comment (wlemkows) [inline doc/vmemcache.md]: .

File: doc/vmemcache.md [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+check *vsize* to tell that apart.
+
+
+int vmemcache_put(VMEMcache *cache,
```
Comment (wlemkows) [inline doc/vmemcache.md]: .

File: doc/vmemcache.md [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+	const void *key, size_t key_size,
+	const void *value, size_t value_size);
+
+Inserts the given key:value pair into the cache.  Returns 0 on success,
+-1 on error.  Inserting a key that already exists will fail with EEXIST.
```
Comment (wlemkows) [inline doc/vmemcache.md]: spaces

File: doc/vmemcache.md [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+# DESCRIPTION #
+
+**libvmemcache** is a volatile key-value store optimized for operating on
+NVDIMM based space, although it can work with any filesystem, be it stored
+in memory (tmpfs) or, less performant, on some kind of a disk.
+
+
+##### Creation #####
+
+VMEMcache *vmemcache_new(const char *path, size_t max_size, size_t segment_size,
```
[the remark this answers is not in the document]
Comment (wlemkows) [reply doc/vmemcache.md]: Or just \*\*vmemcache_new\*\*()

File: doc/vmemcache.md [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+	const void *key, size_t key_size,
+	const void *value, size_t value_size);
+
+Inserts the given key:value pair into the cache.  Returns 0 on success,
+-1 on error.  Inserting a key that already exists will fail with EEXIST.
+
+
+int vmemcache_evict(VMEMcache *cache, const void *key, size_t ksize);
+
+Removes the given key from the cache.  If *key* is null and there is a
```
Comment (wlemkows) [inline doc/vmemcache.md]: spaces

File: doc/vmemcache.md [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+int vmemcache_put(VMEMcache *cache,
+	const void *key, size_t key_size,
+	const void *value, size_t value_size);
+
+Inserts the given key:value pair into the cache.  Returns 0 on success,
+-1 on error.  Inserting a key that already exists will fail with EEXIST.
+
+
+int vmemcache_evict(VMEMcache *cache, const void *key, size_t ksize);
```
Comment (wlemkows) [inline doc/vmemcache.md]: .

File: doc/vmemcache.md [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+Removes the given key from the cache.  If *key* is null and there is a
+replacement policy set, the oldest entry will be removed.  Returns 0 if
```
Comment (wlemkows) [inline doc/vmemcache.md]: .

File: doc/vmemcache.md [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+an entry has been evicted, -1 otherwise.
+
+
+##### Callbacks #####
+
+You can register a hook to be called during eviction or after a cache miss,
+using ``vmemcache_callback_on_evict()`` or ``vmemcache_callback_on_miss()``,
```
Comment (wlemkows) [inline doc/vmemcache.md]: \*\*vmem[]_evict\*\*()

File: doc/vmemcache.md [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+respectively.  The extra *arg* will be passed to your function.
```
Comment (wlemkows) [inline doc/vmemcache.md]: spaces

File: doc/vmemcache.md [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+
+void vmemcache_on_evict(VMEMcache *cache,
```
Comment (wlemkows) [inline doc/vmemcache.md]: .

File: doc/vmemcache.md [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+	const void *key, size_t key_size, void *arg);
+
+Called when an entry is being removed from the cache.  The eviction can't
```
Comment (wlemkows) [inline doc/vmemcache.md]: spaces, spaces (it this by purpose? if not, please fix in in the whole file)

File: doc/vmemcache.md [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+be prevented, but until the callback returns, the entry remains available
+for queries.  The thread that triggered the eviction is blocked in the
+meantime.
+
+
+int vmemcache_on_miss(VMEMcache *cache,
```
Comment (wlemkows) [inline doc/vmemcache.md]: .

File: doc/vmemcache.md [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+	const void *key, size_t key_size, void *arg);
+
+Called when a *get* query fails, to provide an opportunity to insert the
+missing key.  If the callback returns zero, upon return the query will
+be retried (just once -- no call if it fails again).  Note that it's
```
Comment (wlemkows) [inline doc/vmemcache.md]: .

File: doc/vmemcache.md [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+
+
+##### Misc #####
+
+int vmemcache_get_stat(VMEMcache *cache,
+	enum vmemcache_statistic stat,
+	void *value, size_t value_size);
+
+Obtains a piece of statistics about the cache.  The *stat* may be:
+ * VMEMCACHE_STAT_PUT
```
Comment (wlemkows) [inline doc/vmemcache.md]: It would be nice to bold those stats.
Use + list:
\+ \*\*VMEM_STAT_GET\*\* etc. (this and previous list)

**Pull request merged**